plugin/pkg/scheduler/algorithm/priorities/util/non_zero.go

```diff
@@ -48,3 +53,68 @@ func GetNonzeroRequests(requests *api.ResourceList) (int64, int64) {
 	}
 	return outMilliCPU, outMemory
 }
+
+// FilterPodsByNameSpaces filters the pods based the given list of namespaces
+func FilterPodsByNameSpaces(names sets.String, pods []*api.Pod) []*api.Pod {
+	if len(pods) == 0 || len(names) == 0 {
+		return pods
+	}
+	result := []*api.Pod{}
+	for _, pod := range pods {
+		if names.Has(pod.Namespace) {
+			result = append(result, pod)
+		}
+	}
+	return result
+}
+
+// GetNamespacesFromPodAffinityTerm returns a set of names
+// according to the namespaces indicated in podAffinityTerm.
+// if the NameSpaces is nil considers the given pod's namespace
+// if the Namespaces is empty list then considers all the namespaces
+func GetNamespacesFromPodAffinityTerm(pod *api.Pod, podAffinityTerm api.PodAffinityTerm) sets.String {
+	names := sets.String{}
+	if podAffinityTerm.Namespaces == nil {
+		names.Insert(pod.Namespace)
+	} else if len(podAffinityTerm.Namespaces) != 0 {
+		for _, nameSpace := range podAffinityTerm.Namespaces {
+			names.Insert(nameSpace.Name)
+		}
+	}
+	return names
+}
+
+// NodeHasTopologyKey checks if nodeA and nodeB have same label value with given topologyKey as label key.
+// If the topologyKey is nil/empty, regard as the two nodes have same topologykey and label value.
+func NodesHaveSameTopologyKey(nodeA *api.Node, nodeB *api.Node, topologyKey string) bool {
+	if len(topologyKey) == 0 {
```

Comment:
Sorry, the design doc under-specified what was intended for empty topologyKey. The use case was specifically for soft anti-affinity, so a user could say "spread across all topologies" without having to know the names of the topologies, and to make it easy for the administrator to change the set of topologies without users having to change their configs. "All topologies" doesn't really make sense for affinity, only for anti-affinity, which is why I suggested in my comment on TopologyKey in types.go that you should have validation reject empty TopologyKey with PodAffinity. (It's fine to accept empty TopologyKey for soft PodAntiAffinity and hard PodAntiAffinity, even though it's really only useful for soft PodAntiAffinity.) 

So you want something like

```
func NodesHaveSameTopologyKeyInternal(nodeA, nodeB, topologyKey) bool {
    return nodeA.Labels != nil && nodeB.Labels != nil && len(nodeA.Labels[topologyKey]) > 0 && nodeA.Labels[topologyKey] == nodeB.Labels[topologyKey]
}

func NodesHaveSameTopologyKey(nodeA, nodeB, topologyKey) bool {
   if len(topologyKey == 0) {
     // assumes this is allowed only for PodAntiAffinity
     return NodesHaveSameTopologyKeyInternal(nodeA, nodeB, unversioned.LabelHostname) ||
NodesHaveSameTopologyKeyInternal(nodeA, nodeB, unversioned.LabelZoneFailureDomain) ||
NodesHaveSameTopologyKeyInternal(nodeA, nodeB, unversioned.LabelZoneRegion)
   } else {
        return NodesHaveSameTopologyKeyInternal(nodeA, nodeB, topologyKey)
   }
}
```

But I would actually suggest a slight variation on that: instead of hardcoding the label keys like I did above (unversioned.LabelHostname, etc.), you should take the set of label keys as a scheduler command-line argument (an array of strings), for example
```
--failure-domains=kubernetes.io/hostname,failure-domain.beta.kubernetes.io/zone
```

If the flag is not set, then it should default to the three constant values that I listed above (unversioned.LabelHostname, etc). You should also add a comment to pkg/api/unversioned/well_known_labels.go that says something like "if you add a new topology domain here, also consider adding it to the set of default values for the scheduler's --failure-domain command-line argument."

(BTW the reason the values are ORed is because if for some reason a node is missing a label for one topology, you still want the anti-affinity to be true as long as at least one of the labels matches. If we were also going to handle regular affinity, then we would AND for regular affinity and OR for anti-affinity.)